### PR TITLE
feat(Core/SmartAI): Implement SMART_ACTION_FOLLOW_GROUP

### PR DESCRIPTION
**Pull request by Gultask**
## Changes Proposed:
This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

If your pull request promotes complex changes that require a detailed explanation, please describe them in detail specifying what your solution is and what is it meant to address.

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [x] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [x] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

Used https://github.com/cmangos/mangos-wotlk/commit/5e451c0a0f5bfda8e5be67626bf94ed6283aa9b4 as help but it's mostly my own research as well and it's not a cherry-pick per se

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

```sql
DELETE FROM `smart_scripts` WHERE (`source_type` = 0 AND `entryorguid` = 448);
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `event_param5`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_param4`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES
(448, 0, 1, 0, 38, 0, 100, 0, 1, 1, 0, 0, 0, 230, 1, 1, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 1 - Follow Type CIRCLE'),
(448, 0, 2, 0, 38, 0, 100, 0, 1, 2, 0, 0, 0, 230, 1, 2, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 2 - Follow Type SEMI CIRCLE BEHIND'),
(448, 0, 3, 0, 38, 0, 100, 0, 1, 3, 0, 0, 0, 230, 1, 3, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 3 - Follow Type SEMI CIRCLE FRONT'),
(448, 0, 4, 0, 38, 0, 100, 0, 1, 4, 0, 0, 0, 230, 1, 4, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 4 - Follow Type LINE'),
(448, 0, 5, 0, 38, 0, 100, 0, 1, 5, 0, 0, 0, 230, 1, 5, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 5 - Follow Type COLUMN'),
(448, 0, 6, 0, 38, 0, 100, 0, 1, 6, 0, 0, 0, 230, 1, 6, 3, 0, 0, 0, 9, 448, 0, 100, 1, 0, 0, 0, 0, 'Hogger - On Data Set 1 6 - Follow Type ANGULAR');
```

1. .go c id 448
2. .npc add temp 448 a few times
3. .npc set data 1 x

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Gultask** [comment]
I'll go more in detail on what's this about, but if you want an example where SAI is needed, then check for example the Goldshire kids, the imps near the entrance to Shadow Labyrinth or Dimensius the All-Devouring quest

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -2891,6 +2891,101 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
             }
             break;
         }
+        case SMART_ACTION_FOLLOW_GROUP:
+        {
+            if (!e.action.followGroup.followState)
+            {
+                for (WorldObject* target : targets)
+                    if (IsUnit(target))
+                        target->ToCreature()->GetMotionMaster()->MoveIdle();
```
Comment (Helias) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2900]:
```suggestion
                for (WorldObject* target : targets)
                {
                    if (IsUnit(target))
                    {
                        target->ToCreature()->GetMotionMaster()->MoveIdle();
                    }
                }
```

Comment (Kitzunu) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2900]: It is not needed. It hasn't been enforced for a very very long time. And following the code style of the majority of the project this is fine

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 2]
```diff
@@ -2891,6 +2891,101 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+                    break;
+                }
+                case FOLLOW_TYPE_SEMI_CIRCLE_BEHIND:
+                {
+                    for (WorldObject* target : targets)
+                    {
+                        if (IsCreature(target))
+                        {
+                            itr++;
+                            target->ToCreature()->GetMotionMaster()->MoveFollow(me, dist, (M_PI / 2.0f) + (M_PI / membCount) * (itr - 1));
```
Comment (Kitzunu) [inline src/server/game/AI/SmartScripts/SmartScript.cpp:2930]: Just to follow codestyle from the rest of the project, move itr++ after MoveFollow(). Then you can remove itr -1 too

Comment (Gultask) [reply src/server/game/AI/SmartScripts/SmartScript.cpp:2930]: ok, I'll initialize it as itr = 1 at the start so I'll be move it after the MoveFollow

**Comment by Kitzunu** [comment]
Ah I only saw (itr -1) thats why I said you could remove it after moving itr++ down. I missed itr % 2 :P

**Comment by Gultask** [comment]
Wait I forgot about that one :d

**Comment by Gultask** [comment]
still works as it did before 👍

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**